File: web/js/src/Chapter.js
/*
 * Copyright (c) 2026 Taner Sener
 *
 * This file is part of FFmpegKitNext.
 *
 * FFmpegKitNext is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * FFmpegKitNext is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General License for more details.
 *
 * You should have received a copy of the GNU Lesser General License
 * along with FFmpegKitNext. If not, see <http://www.gnu.org/licenses/>.
 */

// LEAF MODULE - MUST NOT IMPORT ANYTHING.

export class Chapter {
    static KEY_ID = 'id';
    static KEY_TIME_BASE = 'time_base';
    static KEY_START = 'start';
    static KEY_START_TIME = 'start_time';
    static KEY_END = 'end';
    static KEY_END_TIME = 'end_time';
    static KEY_TAGS = 'tags';

    constructor(data) {
        this._allProperties = data ?? null;
    }

    getId() {
        return this.getNumberProperty(Chapter.KEY_ID);
    }

    getTimeBase() {
        return this.getStringProperty(Chapter.KEY_TIME_BASE);
    }

    getStart() {
        return this.getNumberProperty(Chapter.KEY_START);
    }

    getStartTime() {
        return this.getStringProperty(Chapter.KEY_START_TIME);
    }

    getEnd() {
        return this.getNumberProperty(Chapter.KEY_END);
    }

    getEndTime() {
        return this.getStringProperty(Chapter.KEY_END_TIME);
    }

    getTags() {
        return this.getProperty(Chapter.KEY_TAGS);
    }

    getStringProperty(key) {
        return this.getProperty(key);
    }

    getNumberProperty(key) {
        return this.getProperty(key);
    }

    getProperty(key) {
        if (
            this._allProperties != null &&
            Object.prototype.hasOwnProperty.call(this._allProperties, key)
        ) {
            return this._allProperties[key];
        }
        return null;
    }

    getAllProperties() {
        return this._allProperties;
    }
}

File: web/src/Chapter.cpp
/*
 * Copyright (c) 2022, 2026 Taner Sener
 *
 * This file is part of FFmpegKitNext.
 *
 * FFmpegKitNext is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * FFmpegKitNext is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General License for more details.
 *
 * You should have received a copy of the GNU Lesser General License
 * along with FFmpegKitNext. If not, see <http://www.gnu.org/licenses/>.
 */

#include "Chapter.h"

ffmpegkit::Chapter::Chapter(std::shared_ptr<ffmpegkit::json::Value> chapterValue)
    : _chapterValue{chapterValue} {}

std::shared_ptr<int64_t> ffmpegkit::Chapter::getId() {
    return getNumberProperty(KeyId);
}

std::shared_ptr<std::string> ffmpegkit::Chapter::getTimeBase() {
    return getStringProperty(KeyTimeBase);
}

std::shared_ptr<int64_t> ffmpegkit::Chapter::getStart() {
    return getNumberProperty(KeyStart);
}

std::shared_ptr<std::string> ffmpegkit::Chapter::getStartTime() {
    return getStringProperty(KeyStartTime);
}

std::shared_ptr<int64_t> ffmpegkit::Chapter::getEnd() {
    return getNumberProperty(KeyEnd);
}

std::shared_ptr<std::string> ffmpegkit::Chapter::getEndTime() {
    return getStringProperty(KeyEndTime);
}

std::shared_ptr<ffmpegkit::json::Value> ffmpegkit::Chapter::getTags() {
    return getProperty(KeyTags);
}

std::shared_ptr<std::string>
ffmpegkit::Chapter::getStringProperty(const char *key) {
    if (_chapterValue == nullptr) {
        return nullptr;
    }
    const ffmpegkit::json::Value *property = _chapterValue->find(key);
    if (property == nullptr) {
        return nullptr;
    }
    return property->getString();
}

std::shared_ptr<int64_t>
ffmpegkit::Chapter::getNumberProperty(const char *key) {
    if (_chapterValue == nullptr) {
        return nullptr;
    }
    const ffmpegkit::json::Value *property = _chapterValue->find(key);
    if (property == nullptr) {
        return nullptr;
    }
    return property->getInt();
}

std::shared_ptr<ffmpegkit::json::Value>
ffmpegkit::Chapter::getProperty(const char *key) {
    if (_chapterValue == nullptr) {
        return nullptr;
    }
    const ffmpegkit::json::Value *property = _chapterValue->find(key);
    if (property == nullptr) {
        return nullptr;
    }
    return std::make_shared<ffmpegkit::json::Value>(*property);
}

std::shared_ptr<ffmpegkit::json::Value> ffmpegkit::Chapter::getAllProperties() {
    if (_chapterValue == nullptr) {
        return nullptr;
    }
    return std::make_shared<ffmpegkit::json::Value>(*_chapterValue);
}
